Add high level interface for HPKE

wbl opened the pull request:
Fixes: Issue #20467

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ x] documentation is added or updated
- [ x] tests are added or updated

wbl commented:
One thing to note is I had trouble with using symbolic names for str2suite in the test.

mattcaswell commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,147 @@
+    out[3] = (unsigned char) (enclen>>24);
+    
+    ct = *box+8+enclen;
+    if (!OSSL_HPKE_seal(ctx, ct, &ctlen, NULL, 0, pt, ptlen)) {
```
No possibility to add AAD? Should we allow this?

wbl replied:
libsodium doesn't expose AAD by default, and for a lot of the "I want to encrypt some data in a blob" usecases it isn't needed, so I think we can leave it out.

OTOH it is just two arguments that can be NULL and 0...

mattcaswell commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,147 @@
+    out[4+enclen] = (unsigned char) (ctlen);
+    out[4+enclen+1] = (unsigned char) (ctlen>>8);
+    out[4+enclen+2] = (unsigned char) (ctlen>>16);
+    out[4+enclen+3] = (unsigned char) (ctlen>>24);
```
Have we invented a format for the "box" here? Or is this length prefixed approach in an RFC somewhere? If so can you provide a reference?

wbl replied:
We have invented a format.

mattcaswell commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,147 @@
+    if (!OSSL_HPKE_str2suite(suitename, &suite)) {
+        return 0;
+    }
+    enclen = ((size_t) box[0]) |
```
Perhaps use the safer PACKET API for this kind of stuff?

wbl replied:
Thanks! that is exactly what I was searching for and didn't find.

mattcaswell commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,147 @@
+        return 0;
+    }
+    if (!OSSL_HPKE_str2suite(suitename, &suite)) {
+        return 0;
```
Perhaps we should be raising ERRs for these kind of errors (`ERR_raise`)

mattcaswell commented on test/hpke_test.c:
```diff
@@ -91,6 +91,45 @@ static int cmpkey(const EVP_PKEY *pkey,
+  }
+  ret = 1;
+ end:
+  EVP_PKEY_free(priv);
```
Shouldn't we be freeing `box` and `opened` here?

wbl commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,147 @@
+                          const char *suitename,
+						  const unsigned char *receiver, size_t receiverlen,
+                          const unsigned char *pt, size_t ptlen,
+						  unsigned char **box, size_t *boxlen)
```
Likely tabs

wbl commented:
I think I've addressed all the comments in this revision.

slontis commented on include/openssl/hpke.h:
```diff
@@ -147,5 +147,15 @@ int OSSL_HPKE_str2suite(const char *str, OSSL_HPKE_SUITE *suite);
 size_t OSSL_HPKE_get_public_encap_size(OSSL_HPKE_SUITE suite);
 size_t OSSL_HPKE_get_recommended_ikmelen(OSSL_HPKE_SUITE suite);
+int OSSL_HPKE_BOX_seal(OSSL_LIB_CTX *libctx, const char *propq,
+                          const char *suite,
```
indentation is all over the place here..

slontis commented on include/openssl/hpke.h:
```diff
@@ -147,5 +147,15 @@ int OSSL_HPKE_str2suite(const char *str, OSSL_HPKE_SUITE *suite);
 size_t OSSL_HPKE_get_ciphertext_size(OSSL_HPKE_SUITE suite, size_t clearlen);
 size_t OSSL_HPKE_get_public_encap_size(OSSL_HPKE_SUITE suite);
 size_t OSSL_HPKE_get_recommended_ikmelen(OSSL_HPKE_SUITE suite);
+int OSSL_HPKE_BOX_seal(OSSL_LIB_CTX *libctx, const char *propq,
```
What does BOX actually mean? (What another library does is not part of the answer :))

wbl replied:
The reason for BOX is that this is the best analogy for an integrated AEAD and encapsulation operation. I'm open to alternative names.

slontis replied:
I think if you look at RFC9180 what you have is just the API in 6.1 (I would add the aad as an optional param)..
which is the single shot API?

slontis reviewed (commented):
Why is there a separate file for this?.. There are already multiple layers of calls in the HKPE API, this is just another one.

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,160 @@
+    if (enclen == 0)
+        goto end;
+
+    /* Format: 4 byte length, encapsulation.
```
````
/*
 * Format ....
 ````

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,160 @@
+    WPACKET pkt;
+    int ret = 0;
+
+    if (*box != NULL) {
```
|| boxlen == NULL

slontis replied:
negative tests for NULL will pick up any other checks you may need to do..

slontis commented on test/hpke_test.c:
```diff
@@ -419,6 +419,40 @@ static int x25519kdfsha256_hkdfsha256_aes128gcm_psk_test(void)
+    unsigned char *opened = NULL;
+    size_t openedlen;
+    int ret  = 0;
+
```
Add negative tests (including params that can not be NULL).

slontis commented on test/hpke_test.c:
```diff
@@ -1960,6 +1994,7 @@ int setup_tests(void)
     ADD_TEST(test_hpke_oddcalls);
     ADD_TEST(test_hpke_compressed);
     ADD_TEST(test_hpke_noncereuse);
+	ADD_TEST(test_hpke_highlevel);
```
indenting wrong

wbl commented:
> Why is there a separate file for this?.. There are already multiple layers of calls in the HKPE API, this is just another one.

No particular reason: I felt that as the hpke file had already gotten quite large it would make sense to separate out the higher level API. I don't feel strongly about this point.

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,160 @@
+#include <internal/packet.h>
+#include <openssl/err.h>
+
+int OSSL_HPKE_BOX_seal(OSSL_LIB_CTX *libctx, const char *propq,
```
Maybe put libctx, propq last in the params..
They are kind of optional parameters just used for algorithm fetching.

wbl replied:
That would align them with OSSL_HPKE_CTX_new, which makes sense. However a bunch of other functions seem to take libctx first and propq last, like EVP_CIPHER_fetch, EVP_PKEY_CTX_new_from_name, so I'm not sure if we'd be making an inconsistency worse by doing that. I'm leaning toward aligning with OSSL_HPKE_CTX_new.

slontis replied:
The reason they are like that in the fetch is that libctx is an important parameter...

slontis replied:
The grease API got it wrong though...

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,160 @@
+        goto err;
+
+    if (!WPACKET_start_sub_packet_u32(&pkt))
+        goto err;
```
if you dont need to raise you can just do
````
if (!X
        || !Y
        || !Z
        ....)
    goto err;
   ````

slontis replied:
This applies in a few places in your code.

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,160 @@
+
+    *ptlen = ctlen;
+    *ptx = OPENSSL_malloc(*ptlen);
+    if (*ptx == NULL) {
```
remove {}

t8m commented on doc/man3/OSSL_HPKE_CTX_new.pod:
```diff
@@ -455,6 +467,16 @@ codepoints. Valid (case-insensitive) names are:
 
+=head2 High-level functions
+
+OSSL_HPKE_BOX_seal() calls OSSL_HPKE_encap and OSSL_HPKE_seal, and serializes the
+outputs into a newly allocated buffer, whose address is returned in box and sets boxlen
+accordingly. It is the caller's responsibility to free I<*box>.
+
+OSSL_HPKE_BOX_open() calls OSSL_HPKE_decap and OSSL_HPKE_open, and returns a
+newly allocated copy of the plaintext in ptx, and sets ptlen to the length. It is
+the caller's responsibility to free I<*ptx>.
```
The problem I see with this is that this invents a new format that is undocumented and non-standard. I am slightly worried of the possible consequences. Will we have to support it forever, is it extensible enough, etc.

t8m replied:
Maybe we do not need the format to be formally standardized but at least a precise description of it should be included in this PR so independent implementations are possible.

mattcaswell replied:
I'm wary of introducing something non-standard. Maybe an OTC discussion is in order. In any case I'd suggest we include some kind of "version" specifier in the format so that it would be easy to make changes to the format in the future if we deem it necessary.

slontis replied:
I was thinking the same thing in regards to a version

wbl commented:
I've gone ahead and added the aad arguments, as well as tests and reordered the libctx and propq arguments. I've also modified the functions to either succeed or set the pointers they were called with to NULL to simplify error handling in callers, and fixed some whitespace issues. I also renamed the functions to follow the suggestion from the RFC.

At this point I think the only outstanding issue is @t8m 's concern about the homebrew standard, which I'm not sure a github PR conversation is necessarily the right way to resolve.

t8m commented:
> Looks like the windows buildbot is unhappy with something not in my code

Yes, but other CI failures are relevant.

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,141 @@
+#include <openssl/err.h>
+
+int OSSL_HPKE_seal_base(OSSL_HPKE_SUITE suite,
+						const unsigned char *receiver, size_t receiverlen,
```
indentation is still wonky.. Should look like this in the editor here.. (No tabs please) 
````
int OSSL_HPKE_seal_base(OSSL_HPKE_SUITE suite,
                                             const unsigned char *receiver, size_t receiverlen,
````

levitte replied:
This is caused by tabs.  I suppose that the author has the setting tab=4 columns.  Github shows them with the setting tab=8 columns

wbl replied:
Thank you, that explains the mysteries of the indentation errors I am seeing.

Perhaps CI could include a no-tabs check to make this easier to discover?

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,141 @@
+
+    if (box == NULL || *box != NULL || boxlen == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto end;
```
The first 3 gotos here can just be return 0

levitte replied:
On the other hand, this makes for consistent code in this function

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,141 @@
+    WPACKET_cleanup(&pkt);
+ boxerr:
+    OPENSSL_free(*box);
+	*box = NULL;
```
tabs here again...

slontis reviewed (commented):
It might be advisable to set tabs as spaces in your editor.. You at least need to change the existing tabs you have to spaces and check alignment..

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,141 @@
+}
+
+int OSSL_HPKE_open_base (OSSL_HPKE_SUITE suite,
+						 EVP_PKEY *recippriv,
```
tabs here?

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,141 @@
+    PACKET encpkt;
+    PACKET ctpkt;
+    int ret = 0;
+    if (ptx == NULL || *ptx != NULL || ptlen == NULL) {
```
blankline after variables

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,141 @@
+        return 0;
+    }
+
+	if (!PACKET_buf_init(&pkt, box, boxlen)
```
tabs here>?

slontis commented on crypto/hpke/hpke_highlevel.c:
```diff
@@ -0,0 +1,141 @@
+    goto end;
+err:
+    OPENSSL_free(*ptx);
+	*ptx = NULL;
```
and here..

t8m commented:
**OTC: According to our policy we cannot accept this as library API as it does not implement a standard but would be happy to accept something like this as a demo.**

wbl commented:
@t8m If there was an ID for this that became an RFC would OpenSSL want to implement it? That's important information in the IETF process: having an implementation be interested moves the process along.

t8m commented:
> @t8m If there was an ID for this that became an RFC would OpenSSL want to implement it? That's important information in the IETF process: having an implementation be interested moves the process along.

Yes, if this was standardized (alone or as part of some higher level protocol), it would be acceptable.

slontis commented:
@wbl the ref to demo meant that you could add it into the demos folder.
